class.c

```diff
@@ -2897,14 +2898,38 @@ rb_singleton_class_get(VALUE obj)
     return klass;
 }
 
-VALUE
-rb_singleton_class(VALUE obj)
+static VALUE
+singleton_class_of_ractor_safe(VALUE obj)
 {
     VALUE klass = singleton_class_of(obj);
```

Comment:
It looks like `singleton_class_of` can raise, so maybe we need to push the locking deeper into there (which I think might end up cleaner)?